JOSESwift/Sources/CryptoImplementation/AES.swift

```diff
@@ -97,37 +123,122 @@ internal struct AES {
     ///   - initializationVector: The initial block.
     /// - Returns: The plain text (decrypted cipher text).
     /// - Throws: `AESError` if any error occurs during decryption.
-    static func decrypt(cipherText: Data, with decryptionKey: Data, using algorithm: ContentEncryptionAlgorithm, and initializationVector: Data) throws -> Data {
+    static func decrypt(
+        cipherText: Data,
+        with decryptionKey: Data,
+        using algorithm: ContentEncryptionAlgorithm,
+        and initializationVector: Data
+    ) throws -> Data {
         switch algorithm {
         case .A256CBCHS512, .A128CBCHS256:
             guard algorithm.checkAESKeyLength(for: decryptionKey) else {
                 throw AESError.keyLengthNotSatisfied
             }
 
-            let decrypted = try aes(
+            let decrypted = ccAESCBCCrypt(
                 operation: CCOperation(kCCDecrypt),
                 data: cipherText, key: decryptionKey,
                 algorithm: algorithm.ccAlgorithm,
                 initializationVector: initializationVector,
                 padding: CCOptions(kCCOptionPKCS7Padding)
             )
 
-            guard decrypted.status == UInt32(kCCSuccess) else {
-                throw AESError.decryptingFailed(description: "Decryption failed with CryptoStatus: \(decrypted.status).")
+            guard
+                let plaintext = decrypted.data,
+                decrypted.status == UInt32(kCCSuccess)
+            else {
+                throw AESError.decryptingFailed(description: "Decryption failed with status: \(decrypted.status).")
+            }
+
+            return plaintext
+        }
+    }
+
+    /// Encrypts the given raw key using AES key wrap.
+    /// - Parameters:
+    ///   - rawKey: The raw key to encrypt
+    ///   - keyEncryptionKey: The key used to encypt the raw key
+    ///   - algorithm: The algorithm to use for AES key wrap
+    static func keyWrap(rawKey: Data, keyEncryptionKey: Data, algorithm: KeyManagementAlgorithm) throws -> Data {
+        switch algorithm {
+        case .A128KW, .A192KW, .A256KW:
+            guard
+                let keyLengthOk = algorithm.checkAESKeyLength(for: keyEncryptionKey),
+                keyLengthOk
+            else {
+                throw AESError.keyLengthNotSatisfied
+            }
+
+            // See https://tools.ietf.org/html/rfc3394#section-2.2.3.1
+            // The default iv is defined to be the hexadecimal constant A6A6A6A6A6A6A6A6
+            let iv = Data(bytes: CCrfc3394_iv, count: CCrfc3394_ivLen)
+
+            let wrapped = ccAESKeyWrap(rawKey: rawKey, keyEncryptionKey: keyEncryptionKey, iv: iv)
+
+            guard
+                let wrappedKey = wrapped.data,
```

Comment:
Again, no need for binding here.
```swift 
wrapped.data?.status == kCCSuccess
```

same applies below 👇

Reply:
In this case I reuse the let-bound `wrappedKey` later though.

Reply:
Aha, my mistake then.